### ISSUE TITLE
Cannot reproduce the accuracy of llama2-7b on CS using the llama2_7B_DoRA.sh script.

### ISSUE THREAD

**Issue by LuletterSoul** [NVlabs/DoRA#9]
Thank you to the author for making their outstanding work open source. The improvements compared to LoRA are truly exciting! I am currently trying to reproduce your results, but I have encountered some issues. This my reproduced results:
```
PEFT Method    DoRA（bs=16）

BoolQ          0.72
PIQA           0.81
SIQA           0.78
HellaSwag      0.85
WinoGrande     0.80
ARC-e          0.80
ARC-c          0.65
OBQA           0.79
Avg.           0.77
```
The accuracy is much lower than the one given in `ReadMe.MD`. Where is the problem?

**Comment by nbasyl** [comment]
Hi, thanks for your interest in our work, have you tried using our provided DoRA weight to reproduce the number?

**Comment by LuletterSoul** [comment]
> Hi, thanks for your interest in our work, have you tried using our provided DoRA weight to reproduce the number?

No, I plan to use the official llama2-7b weights for fine-tuning, but I am unable to reproduce the results you provided. Additionally, I discovered that：

1. The adapter used in `llama2_7B_DoRA.sh` is actually LoRA, so I modified it to DoRA. 
2. `sh llama_7B_Dora_eval.sh ./finetuned_result/dora_r32 ./finetuned_result/dora_r32 0` command should be `sh llama_7B_Dora_eval.sh  ./finetuned_result/dora_r32 0`. Otherwise, the inference program will not be able to detect the GPU.

Did I overlook some details that led to inaccurate reproduction? If you could provide some specifics, that would be helpful.

**Comment by nbasyl** [comment]
Yeah, I think this is where the problem is, thanks for pointing those typos out, will update them.

**Comment by LuletterSoul** [comment]
Just now, I was able to reproduce the accuracy of your table using the DORA Weight you provided.

**Comment by LuletterSoul** [comment]
> Yeah, I think this is where the problem is, thanks for pointing those typos out, will update them.

In fact, the results in the table were obtained using a modified version of DoRA, which is very strange as they differ greatly from the results you provided.

**Comment by nbasyl** [comment]
what do you mean modified version of DoRA, I don't quite get it

**Comment by nbasyl** [comment]
Are you referring to your table? In that case, those results are the LoRA results which align with our LoRA avg scores as well

**Comment by LuletterSoul** [comment]
> what do you mean modified version of DoRA, I don't quite get it

I apologize for the confusion. I mean that I used the corrected version of llama2_7B_DoRA.sh to obtain the results shown in my table, as I discovered the typos while attempting to reproduce the results.

In addition, could you share the loss records for fine-tuning the Llama2-7B model using the DoRA? I would to compare the losses with yours.

**Comment by nbasyl** [comment]
[trainer_state.json](https://github.com/NVlabs/DoRA/files/15466349/trainer_state.json)

**Comment by nbasyl** [comment]
So your problem now is that even with the correct sh file, you can not reproduce the result?

**Comment by LuletterSoul** [comment]
> So your problem now is that even with the correct sh file, you can not reproduce the result?

That'is right.

**Comment by nbasyl** [comment]
This is pretty strange, can you also try llama3, and let me know?

**Comment by LuletterSoul** [comment]
> This is pretty strange, can you also try llama3, and let me know?
I just compared the losses you provided, and I found that my eval loss is smaller (around 0.65). I'm not sure if this indicates that the model is overfitting. Since a single GPU doesn't have enough memory, I used 8 GPUs for parallel training. At the same time, I set the batch_size of each GPU to 2, keeping the total batch_size at 16.

I cannot download llama3 on my computer, so I haven't had a chance to try it out yet.

**Comment by nbasyl** [comment]
ohh, now I know why since my code is based on https://github.com/AGI-Edgerunners/LLM-Adapters, although theoretically speaking, it should also support multi-gpu training, but apparently, many people have been reporting issues regarding multi-gpu training, so I think that is why you could not reproduce the result. 

But, I have tested the code on a single 3090 by setting smaller --micro_batch_size, and got similar results. Is you gpu even smaller than 3090?

You can also try FSDP/QDoRA which supports multi-gpu training as well as model sharding, I have provided step-by-step instructions under the folder QDoRA, feel free to give it a try.

**Comment by LuletterSoul** [comment]
My GPU is an A100 with 40960MiB of memory. I also tried setting the micro_batch_size to 8, which allowed for single-card training, but this configuration still could not produce the results shown in your table. This is why I need to use 8-card training, as it aligns with your configuration, which has a batch size of 16 per forward pass.

**Comment by nbasyl** [comment]
Got it, I will investigate these issues! and will let you know if I have any updates

**Comment by nbasyl** [comment]
@LuletterSoul, Hi, I just reran the DoRA experiments on llama2, and here are the results

![image](https://github.com/NVlabs/DoRA/assets/45586614/7b4a61e0-0348-4ee7-874b-777918ddaa18)

you can check the corresponding Dora weights here: https://drive.google.com/drive/folders/1lfUi7tEX3sgKDHeVWncFxykMzXfrn3bR?usp=sharing

Indeed the r=32 is not as good as my first run, which might be due to potential overfitting (lower loss than the first run), and the r=16 is roughly the same as the first run. 

There is a recent paper, that can also reproduce our results, you can check it here: https://arxiv.org/pdf/2405.15525v1, it also shows that our method can significantly outperform LoRA on lower-rank settings, so I assume that since DoRA improves the learning capability of LoRA, it also becomes easier to overfit on larger rank, which also explain why lower rank DoRA is more stable.

**Comment by azuryl** [comment]
> @LuletterSoul, Hi, I just reran the DoRA experiments on llama2, and here are the results
> 
> ![image](https://private-user-images.githubusercontent.com/45586614/334813870-7b4a61e0-0348-4ee7-874b-777918ddaa18.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1MTk4NTMsIm5iZiI6MTcyNDUxOTU1MywicGF0aCI6Ii80NTU4NjYxNC8zMzQ4MTM4NzAtN2I0YTYxZTAtMDM0OC00ZWU3LTg3NGItNzc3OTE4ZGRhYTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI0VDE3MTIzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxNGQxYjI0OWRlZjUwODBkNDRmYTM4MjRkYTQzMjgxMWNmZTlkMmNjNDM0MDVjZTYzOTM3MmMzOWZjODZhYjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Y7yaDgGMulGEiUgQQfE7BvQf6OwiNMhOc0S3fCuYaBw)
> 
> you can check the corresponding Dora weights here: https://drive.google.com/drive/folders/1lfUi7tEX3sgKDHeVWncFxykMzXfrn3bR?usp=sharing
> 
> Indeed the r=32 is not as good as my first run, which might be due to potential overfitting (lower loss than the first run), and the r=16 is roughly the same as the first run.
> 
> There is a recent paper, that can also reproduce our results, you can check it here: https://arxiv.org/pdf/2405.15525v1, it also shows that our method can significantly outperform LoRA on lower-rank settings, so I assume that since DoRA improves the learning capability of LoRA, it also becomes easier to overfit on larger rank, which also explain why lower rank DoRA is more stable.

the link can not open
